substrate/client/rpc-spec-v2/src/chain_head/chain_head_follow.rs

```diff
@@ -705,71 +710,109 @@ where
+		let tx_send = Arc::new(parking_lot::Mutex::new(tx_send));
+		let mut events = move |event| {
+			match event {
 				NotificationType::InitialEvents(events) => Ok(events),
 				NotificationType::NewBlock(notification) =>
 					self.handle_import_blocks(notification, &startup_point),
 				NotificationType::Finalized(notification) =>
 					self.handle_finalized_blocks(notification, &startup_point),
 				NotificationType::MethodResponse(notification) => Ok(vec![notification]),
-			};
-
-			let events = match events {
-				Ok(events) => events,
-				Err(err) => {
-					debug!(
-						target: LOG_TARGET,
-						"[follow][id={:?}] Failed to handle stream notification {:?}",
-						self.sub_id,
-						err
-					);
-					_ = sink.send(&FollowEvent::<String>::Stop).await;
-					return Err(err)
-				},
-			};
+			}
+			.map_err(|err| {
+				debug!(
+					target: LOG_TARGET,
+					"[follow][id={:?}] Failed to handle stream notification {:?}",
+					self.sub_id,
+					err
+				);
+				err
+			})
+		};
 
-			for event in events {
-				if let Err(err) = sink.send(&event).await {
-					// Failed to submit event.
-					debug!(
-						target: LOG_TARGET,
-						"[follow][id={:?}] Failed to send event {:?}", self.sub_id, err
-					);
-
-					let _ = sink.send(&FollowEvent::<String>::Stop).await;
-					// No need to propagate this error further, the client disconnected.
-					return Ok(())
+		let stream = stream
+			.map(move |event| events(event))
+			.then(move |event| {
+				let handle = handle.clone();
+				let tx_send = tx_send.clone();
+				async move {
+					let handle = handle.clone();
+					let tx_send = tx_send.clone();
+
+					let result = match event {
+						Ok(events) => stream::iter(events),
+						Err(err) => {
+							{
+								// TODO: this is very hacky
+								// grab the lock
+								// swap the channels
+								// return the error to the main channel
+								let mut tx_send = tx_send.lock();
+								let (mut s, _receiver) =
+									oneshot::channel::<SubscriptionManagementError>();
+								mem::swap(&mut s, &mut tx_send);
+								s.send(err).expect("shouldn't happpen.");
+							}
+							handle.abort();
+							stream::iter(vec![])
+						},
+					};
+					result
 				}
-			}
+			})
+			.flatten();
 
-			stream_item = stream.next();
-			stop_event = next_stop_event;
-		}
+		tokio::pin!(stream);
+
+		let connection_closed = sink.closed().fuse();
+		tokio::pin!(connection_closed);
+
+		let sink_future = sink
+			.pipe_from_stream(stream, sc_rpc::utils::BoundedVecDeque::new(MAX_BUFFERED_EVENTS))
+			.fuse();
+
+		tokio::pin!(sink_future);
+		tokio::pin!(rx_stop);
 
-		// If we got here either:
-		// - the substrate streams have closed
-		// - the `Stop` receiver was triggered internally (cannot hold the pinned block guarantee)
-		// - the client disconnected.
+		let result = futures::select! {
+			err =  tx_receive => {
+				Err(err.expect("should not happen.qed"))
+			}
+			_ = connection_closed => {
+				cancel_handle.abort();
```

Comment:
can we just abort regardless of what stopped it?

```rust
let result = futures::select! {
  _ = rx_stop => Ok(()),
 _ = sink_future => Ok(()),
 // ...
};
cancel_handle.abort();
```

Reply:
done